uswds/uswds
# USWDS - Prose: Placeholder fix. Resolves #5094

Alternative solution from #4962 

## Summary
Reverts `usa-prose` placeholders to use typeset mixins to resolve compilation errors. 

## Breaking change
:warning: This is potentially a breaking change. This PR removes `usa-prose-` mixins and placeholders. Users should instead use `typeset-` mixins in their place. 

## Site changelog PR

https://github.com/uswds/uswds-site/pull/2095

## Related issue
Closes #4962

## Preview link
[Prose →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-usa-prose-selector-fix/iframe.html?args=&id=components-prose--default&viewMode=story)

## Problem statement
This work resolves two separate issues:
1. Build errors while extending `%usa-prose-` placeholders.
2. Compilation errors with mixins nested under the `& >` selector

### 1: Build errors with placeholder:

Sass can throw errors when we rely on `@extend` and placeholders in certain situations.
From https://github.com/uswds/uswds/issues/4962: 
> I’m trying to configure USWDS inside an @layer rule using the load-css function from sass:meta. When I try to @use "uswds-core" in my own Sass modules so that I can access the tokens for my styles, I get a target selector not found error for %usa-prose-p.


### 2: Compilation errors with mixins nested under the `& >` selector

This would result in the unexpected assignment of the `&` selector and in turn, incorrectly attach styles to classes

## Solution
1. Revert `usa-prose` placeholder to a mixin
2. Manually paragraph and heading styles from the `typeset` mixins and directly into the `usa-prose` style definitions
3. Removed redundant and unused `usa-prose` mixins in favor of `typeset-` mixins

From @amyleadem in #5043:
Sass seems to get confused when * + & is brought in from another file via a mixin (more details in https://github.com/uswds/uswds/pull/4576). Additionally, Sass can throw errors in limited circumstances when we rely on @extend and placeholders.

The key conclusion is that `* + &` does not currently seem to be a viable addition to mixins or placeholders. By moving all `* + &` out of mixins and placeholders and directly into the style definitions, we can remove the risk of Sass confusion.

Note: The downside of this solution is that it requires some code repetition, but that might be slightly mitigated by creating variables/settings for the heading and paragraph top margins.

## Major changes
This PR removes `usa-prose-` mixins and placeholders. Users should instead use `typeset-` mixins in their place. 

### List of mixins removed:
- usa-prose-link
- usa-prose-display
- usa-prose-p
- usa-prose-heading
- usa-prose-_h1 through h6_

## New settings variables
- `$theme-heading-margin-top` 
- `$theme-paragraph-margin-top`

## Testing and review
1. Confirm that Sass compiles without error
2. Visit [Prose Storybook](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-usa-prose-selector-fix/iframe.html?args=&id=components-prose--default&viewMode=story)
3. Inspect `usa-prose` styles and make sure `* + &` selectors are correctly targetting headings and p tags
4. Use `npm link` to create a symlink to the package
5. Clone this [test repo](https://github.com/esheehan-gsl/uswds-layer-test)
6. Use `npm link @uswds/uswds` to replace USWDS package with the symlink
7. Run `npm build` and confirm no build errors
